### PR TITLE
Internationalization Guide with a Translator class

### PR DESCRIPTION
**Pull request by btihen**
### **Internationalization using an class approach**

**Purpose**: 
This PR documents the **class** approach to Internationalization

**Approach**: 
This PR is a competitor to https://github.com/luckyframework/website/pull/209 - which uses the **module** approach to Internationalization.  It is assumed only one or the other PR will be chosen and posted.

### CODE REVIEW

**Comment by btihen** [comment]
I am not sure how to tell md that the code shouldn't be expanded - do I need to escape somehow.

```
web    | In src/actions/guides/frontend/internationalization.cr:190:15
web    | 
web    |  190 | h3 "#{I18n.t("me.email", @translator.lang)}:  #{@current_user.email}"
web    |              ^---
web    | Error: undefined constant I18n
```

File: src/actions/guides/frontend/internationalization.cr [hunk 1]
```diff
@@ -0,0 +1,341 @@
+class Guides::Frontend::Internationalization < GuideAction
+  guide_route "/frontend/internationalization"
+
+  def self.title
+    "Internationalization - Handle Multiple Languages"
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]: This wraps oddly in the sidebar so I think just "Internationalization" is ok

File: src/actions/guides/frontend/internationalization.cr [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+  end
+
+  def markdown : String
+    <<-MD
+    This describes how to handle USERS with multiple language preferences.  At somepoint we can document how to add dynamic switching from the frontend too.
+
+    ## **Step 1** - add i18n shard to `shard.yml`
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]: The ** is not rendered well in the table of contents. I think the bold can be removed. Also could you capitalize the sentences after the `-`

File: src/actions/guides/frontend/internationalization.cr [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        auth_guides: "Check out the authentication guides"
+        modify_page: "Modify this page"
+        after_signin: "Change where you go after sign in"
+    ```
+
+    ## **Step 3** - configure i18n to work with Lucky
+    ```
+    # config/i18n.cr
+    require "i18n"
+    require "i18n/backends/yaml"
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]: I think these `requires` should actually go in Step 1 and should be required in `src/shards.cr`.

File: src/actions/guides/frontend/internationalization.cr [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+      I18n.backend = I18n::Backends::YAML.new.tap do |backend|
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]:
```suggestion
    I18n.backend = I18n::Backends::YAML.new.tap do |backend|
```

File: src/actions/guides/frontend/internationalization.cr [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+      backend.load_paths << Dir.current + "/config/locales"
+      backend.load
+    end
+    ```
+
+    ## **Step 4** - create a user migration
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]:
```suggestion
    ## Step 4 - Add 'lang' to users table
```

File: src/actions/guides/frontend/internationalization.cr [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+      end
+    end
+    ```
+
+    and of course migrate
+    ```
+    lucky db.migrate
+    ```
+
+    ## **Step 5** - update the user model (add the new attribute)
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]:
```suggestion
    ## Step 5 - Add lang column to User model
```

File: src/actions/guides/frontend/internationalization.cr [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    ```
+    # src/models/user.cr
+      table do
+        column lang : String
+        column email : String
+        column encrypted_password : String
+       end
+    ```
+
+    ## **Step 6** - Update sign_up_user.cr (to validate the available languages) - ideally at somepoint in the future - lucky will autodect and autocreate the language list based on the config/locales yml files.
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]: I agree Lucky could autoload this but I think can be removed from the headline

File: src/actions/guides/frontend/internationalization.cr [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    # src/components/translator.cr
+    class Translator
+      getter lang : String
+
+      def initialize(user : User? = nil)
+        @lang = user.try(&.lang) || "en"
+      end
+
+      def t(key : String)
+        I18n.t(key, @lang) # this seems better but not available
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]: What does this mean "better but not available"? I'm not sure I understand

File: src/handlers/legacy_redirect_handler.cr [hunk 9]
```diff
@@ -12,6 +12,7 @@ class LegacyRedirectHandler
     "/guides/custom-queries"                      => Guides::Database::RawSql,
     "/guides/saving-with-forms"                   => Guides::Database::ValidatingSaving,
     "/guides/asset-handling"                      => Guides::Frontend::AssetHandling,
+    "/guides/internationalization"                => Guides::Frontend::Internationalization,
```
Comment (paulcsmith) [inline src/handlers/legacy_redirect_handler.cr]: Since this is a new page, it doesn't need a legacy redirect. New pages go in `src/models/guides_list.cr`. This one should go last in the FrontEnd category

File: src/actions/guides/frontend/internationalization.cr [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+    ```
+
+    ```
+    # src/pages/sign_ups/new_page.cr
+    class SignUps::NewPage < AuthLayout
+      needs operation : SignUpUser
+
+      def content
+        h1 I18n.t("auth.sign_up", @translator)
+        render_sign_up_form(@operation)
```
Comment (paulcsmith) [inline src/actions/guides/frontend/internationalization.cr]: I think these could be clarified by removing most of the code that doesn't need to be changed. For example: 

```crystal
    class SignUps::NewPage < AuthLayout
      def content
        h1 I18n.t("auth.sign_up", @translator)
        # ...
      end

      private def render_sign_up_form(op)
        form_for SignUps::Create do
          # ...
          submit I18n.t("auth.sign_up", @translator), flow_id: "sign-up-button"
        end
        link I18n.t("auth.sign_in", @translator), to: SignIns::New
      end

      # ...
    end
```

**Review by paulcsmith — COMMENTED** [review]
Left a bit of feedback. Will take another look again later with a fresh set of eyes. Overall this is looking good. Thanks so much for starting it

File: src/actions/guides/frontend/internationalization.cr (hunk 3 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: cool learning a lot - thanks

File: src/actions/guides/frontend/internationalization.cr (hunk 5 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: oh yeah much better

File: src/actions/guides/frontend/internationalization.cr (hunk 8 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: oops - spurious reminders that should be deletee

File: src/handlers/legacy_redirect_handler.cr (hunk 9 above)
Comment (btihen) [reply src/handlers/legacy_redirect_handler.cr]: ah yup. now it does what I expect

File: src/actions/guides/frontend/internationalization.cr (hunk 10 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: OK - I was thinking then the person could copy and paste.  But this clarifies what to do better.  I have tried to cleanup - but we can make further cleanups too.

**Comment by paulcsmith** [comment]
I also deployed a review app so whenever you make updates to this PR you can see what it looks like here: https://lucky-websit-internatio-esmfcx.herokuapp.com/guides/frontend/internationalization

**Comment by btihen** [comment]
OK - should have most comments addressed now

File: src/actions/guides/frontend/internationalization.cr (hunk 1 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: yup fixed - I didn't see it since I had the link wrong (that's fixed too)

File: src/actions/guides/frontend/internationalization.cr (hunk 2 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: fixed

File: src/actions/guides/frontend/internationalization.cr (hunk 4 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: yup fixed

File: src/actions/guides/frontend/internationalization.cr (hunk 5 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: fixed & caps too

File: src/actions/guides/frontend/internationalization.cr (hunk 6 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: fixed - much better

File: src/actions/guides/frontend/internationalization.cr (hunk 7 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: removed from headline - if you have tips on how to easily do that and add it to the validation and user drop-down - that would be a cool improvement to the doc

File: src/actions/guides/frontend/internationalization.cr (hunk 8 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: fixed

**Comment by btihen** [comment]
> I also deployed a review app so whenever you make updates to this PR you can see what it looks like here: https://lucky-websit-internatio-esmfcx.herokuapp.com/guides/frontend/internationalization

That's pretty cool.  I got it running locally too - so I can see changes before they get pushed.  At first I thought the .env errors were a show stopper - but then I realized it still worked - will it auto-push the next push in the branch?

**Comment by paulcsmith** [comment]
Looks like I need to trigger a manual refresh for the review app. I think because you are not an admin/owner on the repo I have to manually do it 🤔

**Comment by btihen** [comment]
Bummer - sorry about that.  Thanks for accepting this guide.  I think it will be helpful to those in multi-lingual countries!

**Comment by btihen** [comment]
In the other PR https://github.com/luckyframework/website/pull/209 - I have all the features working.  If we choose class over module approach - I will comeback to this branch and finish it - I have both test branches of code in my sample project.  This got complex - probably tests would also be a good idea.  I didn't realize how many edge cases and places we would need to translate.

File: src/actions/guides/frontend/internationalization.cr [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+      me:
+        email: Mail
+        profile: Profil
+        next: "Zunächst bitte kontrolliere"
+        auth_guides: "Bitte kontrolliere die 'Authentication Guides'"
+        modify_page: "Um diese Seite zu Ändern"
+        after_signin: "Eine andere Seite nach anmelden"
+    ```
+
+    **NOTE:** All lang yml files need all the same keys defined. I18n shard is pretty finicky and its error messages aren't very helpful _(expect a little frustation getting the ymls correct and debugged)._
```
Comment (jwoertink) [inline src/actions/guides/frontend/internationalization.cr]:
```suggestion
   > Be sure that all lang yml files contain the same keys. If there's no translation for that key, just leave the value blank.
```

For these type of notes, we generally use the blockquotes. I'd also like to keep the connotation of the message on a more positive note. So I'm not completely sold on the suggested wording, but hopefully that gives you an idea of the direction. Thanks!

File: src/actions/guides/frontend/internationalization.cr [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+
+    ## Step 3 - Configure i18n within Lucky
+    ```
+    # config/i18n.cr
+    I18n.backend = I18n::Backends::YAML.new.tap do |backend|
+      backend.load_paths << Dir.current + "/config/locales"
```
Comment (jwoertink) [inline src/actions/guides/frontend/internationalization.cr:125]: Out of curiosity, would this still work  with `backend.load_paths << "config/locales"`? Not a huge deal, just wondering.

File: src/actions/guides/frontend/internationalization.cr [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+        # ...
+      end
+      # ...
+    end
+    ```
+
+    ## Step 6 - Create a Translator class
+
+    TODO: finalize the best approach & ideally be ready to capture any Frontend JS overrides
+    TODO: autodect and autocreate the language list based on the config/locales yml files.
```
Comment (jwoertink) [inline src/actions/guides/frontend/internationalization.cr]:
```suggestion
    TODO: autodetect and autocreate the language list based on the config/locales yml files.
```

File: src/actions/guides/frontend/internationalization.cr [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+
+    ## Step 8 - Add language to signup form
+
+    TODO: use the same list found in validations (shared constant?)
+    ```
+    # comming soon - when I figure out dropdowns
+    ```
+
+
+    ## Step 8 - Internationalize pages
```
Comment (jwoertink) [inline src/actions/guides/frontend/internationalization.cr]: Step 9?

File: src/actions/guides/frontend/internationalization.cr [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+          # ...
+          submit I18n.t("auth.sign_up", @translator), flow_id: "sign-up-button"
+        end
+        link I18n.t("auth.sign_in", @translator), to: SignIns::New
+      end
+      # ...
+    end
+    ```
+
+    _Thanks to @paulcsmith & @bdtomlin and others for helping and guiding this document._
```
Comment (jwoertink) [inline src/actions/guides/frontend/internationalization.cr]: I agree, these are some awesome people; however, we don't really credit any on any of the other guides. My recommendation would be to omit this, but I'll leave that final word to @paulcsmith  😄

**Review by jwoertink — CHANGES_REQUESTED** [review]
This is awesome! Thanks for taking this on. I know lots of people have requested it. I just made a few small suggestions. I'll take a look again later. It looks like you may still be filling in a few things, so that's fine.

File: src/actions/guides/frontend/internationalization.cr (hunk 11 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: Thanks for the suggestion. Updated

File: src/actions/guides/frontend/internationalization.cr (hunk 13 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: I removed after I realized that the laguages need to be named in the selector & I don't see how that can be automated.

File: src/actions/guides/frontend/internationalization.cr (hunk 14 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: fixed - thanks

File: src/actions/guides/frontend/internationalization.cr (hunk 15 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: I removed - but glad too add it back too

Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr]: @jwoertink - can you also look at the modular approach and see what approach you think is better (class or module) at the PR: https://github.com/luckyframework/website/pull/209

Thanks for the review

**Comment by btihen** [comment]
> This is awesome! Thanks for taking this on. I know lots of people have requested it. I just made a few small suggestions. I'll take a look again later. It looks like you may still be filling in a few things, so that's fine.

indeed - I can go back and finish the doc for this approach.  @paulcsmith  - also suggested trying out a module Translate -- which has some interesting properties - so I have documented both approaches and we can choose one or the other or both if that seems like the best approach - but I asuume there is a standard approach to new features (I'm just not yet familiar enough with the code yet).

**Comment by jwoertink** [comment]
Ah, I see now. I'm probably more partial to the module approach in this instance because my apps don't have a `User` model. We have a `Member` model instead, and our language translation stuff is done by passing a query param in the URL.

I haven't tested this, but I *think*, we could do this with the modules too.

```crystal
class MyTranslator
  include Translator
  #...
end

class SomePage < MainLayout
  needs translator : Translator
end

html(SomePage, translator: MyTranslator.new)
```

My guess is that if you include the module, then when it looks for that type, any class that includes the module would still match.

File: src/actions/guides/frontend/internationalization.cr (hunk 12 above)
Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr:125]: I'll try it out.  I just went with the closest thing to the i18n documentation

Comment (btihen) [reply src/actions/guides/frontend/internationalization.cr:125]: > Out of curiosity, would this still work with `backend.load_paths << "config/locales"`? Not a huge deal, just wondering.

Nope  - it compiles just fine and then crashes on the first translation it tries.

**Comment by btihen** [comment]
> Ah, I see now. I'm probably more partial to the module approach in this instance because my apps don't have a `User` model. We have a `Member` model instead, and our language translation stuff is done by passing a query param in the URL.
> 
> I haven't tested this, but I _think_, we could do this with the modules too.
> 
> ```crystal
> class MyTranslator
>   include Translator
>   #...
> end
> 
> class SomePage < MainLayout
>   needs translator : Translator
> end
> 
> html(SomePage, translator: MyTranslator.new)
> ```
> 
> My guess is that if you include the module, then when it looks for that type, any class that includes the module would still match.

Interesting idea.  I will leave the two docs as they are for now and if we choose a mixed approach then I can document that. (It took me longer than expected to do this in different ways) - but it seems it a good way to learn about lucky.

**Comment by btihen** [comment]
Closining this PR (Class Based approach in favor of the Method Based Approach - it seems we (all 3 of us agree).
